### ISSUE TITLE
maybe suggest running `cargo update` if there's a previously selected version conflict

### ISSUE THREAD

**Issue by mozkeeler** [rust-lang/cargo#5484]
I added a new dependency to a project, ran `cargo build` and got this:

```
    error: failed to select a version for `libc`
    all possible versions conflict with previously selected versions of `libc`
    required by package `dlopen_derive v0.1.0`
        ... which is depended on by `dlopen v0.1.0`
        ... which is depended on by `ooppkcs11rs v0.0.1 (file:///home/keeler/src/ooppkcs11rs)`
      previously selected package `libc v0.2.23`
        ... which is depended on by `ipc-channel v0.10.0`
        ... which is depended on by `ooppkcs11rs v0.0.1 (file:///home/keeler/src/ooppkcs11rs)`
      possible versions to select: 0.2.40, 0.2.39, 0.2.38, 0.2.37, 0.2.36, 0.2.35, 0.2.34, 0.2.33, 0.2.32, 0.2.31, 0.2.30, 0.2.29 
```

My assumption was this was because ipc-channel didn't support libc >0.2.23, but there's no such restriction in ipc-channel's Cargo.toml, so this was just confusing. I ended up removing Cargo.lock and rebuilding, and this worked. A colleague later told me `cargo update` is the real way to fix this.

**Comment by alexcrichton** [comment]
cc @Eh2406

**Comment by Eh2406** [comment]
Ya we can definitely do better. @alexcrichton Is there an easy way to tell if a dependency is pind by a lock file, if so we should definitely add it to the output. Ore is there some way to trigger a thing like we do for https://github.com/rust-lang/cargo/pull/5180 ?

**Comment by alexcrichton** [comment]
@Eh2406 yeah you can use [`precise`](https://github.com/rust-lang/cargo/blob/5986492773e6de61cced57f457da3700607f4a3a/src/cargo/core/source/source_id.rs#L273) in general and if it's `Some` the dependency likely came from a lock file

**Comment by Eh2406** [comment]
I added a note reminding myself to try that in our rework of the wording of these are messages.

On a different note, isn't this exactly the situation #5180 what's supposed to solve? If not how hard would it be to expand that to work here?

@mozkeeler by the way, what version of cargo gave you that message?

**Comment by mozkeeler** [comment]
@Eh2406 well, the version I have now is `cargo 0.26.0 (41480f5cc 2018-02-26)` but I may have updated it as part of trying to figure this out... (although I don't think even doing that helped until I removed Cargo.lock...)

**Comment by Eh2406** [comment]
@mozkeeler Thanks for the info! That explains it, #5180 merged after [0.26.0](https://github.com/rust-lang/cargo/tree/0.26.0) branched. Sorry for the noize.

**Comment by Ekleog** [comment]
FWIW, this is something I just bumped into. Even without checking whether the dependency comes from `Cargo.lock`, maybe just adding “sometimes this error can be fixed by running `cargo update`” would be a good first step?
